core/src/main/java/org/apache/calcite/rel/rules/AggregateJoinJoinRemoveRule.java

```diff
@@ -0,0 +1,129 @@
+ * sales as s
+ * left join product_class pc
+ * on s.product_id = pc.product_id</pre></blockquote>
+ *
+ */
+public class AggregateJoinJoinRemoveRule extends RelOptRule {
+  public static final AggregateJoinJoinRemoveRule INSTANCE
+      = new AggregateJoinJoinRemoveRule(LogicalAggregate.class,
+      LogicalJoin.class, RelFactories.LOGICAL_BUILDER);
+
+  /** Creates an AggregateJoinJoinRemoveRule. */
+  public AggregateJoinJoinRemoveRule(
+      Class<? extends Aggregate> aggregateClass,
+      Class<? extends Join> joinClass, RelBuilderFactory relBuilderFactory) {
+    super(
+        operand(aggregateClass,
+            operandJ(joinClass, null,
+                join -> join.getJoinType() == JoinRelType.LEFT,
+                operandJ(joinClass, null,
+                    join -> join.getJoinType() == JoinRelType.LEFT, any()))),
+        relBuilderFactory, null);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Aggregate aggregate = call.rel(0);
+    final Join topJoin = call.rel(1);
+    final Join bottomJoin = call.rel(2);
+    int leftBottomChildSize = bottomJoin.getLeft().getRowType()
+        .getFieldCount();
+
+    // Check whether the aggregate uses unexpected columns.
+    if (!RelOptUtil.removeJoin(aggregate, leftBottomChildSize,
+        bottomJoin.getRowType().getFieldCount())) {
+      return;
+    }
+
+    // Check whether the top join uses unexpected columns.
+    final List<Integer> leftKeys = new ArrayList<>();
+    RelOptUtil.splitJoinCondition(topJoin.getLeft(), topJoin.getRight(),
+        topJoin.getCondition(), leftKeys, new ArrayList<>(),
+        new ArrayList<>());
+    if (leftKeys.stream().filter(s -> s >= leftBottomChildSize).count() > 0) {
+      return;
+    }
+
+    // Check whether left join keys in top join and bottom join are equal.
+    final List<Integer> leftChildKeys = new ArrayList<>();
+    RelOptUtil.splitJoinCondition(bottomJoin.getLeft(), bottomJoin.getRight(),
+        bottomJoin.getCondition(), leftChildKeys, new ArrayList<>(),
+        new ArrayList<>());
+    if (!leftKeys.equals(leftChildKeys)) {
+      return;
+    }
+
+    // Adjust group keys according to offset.
+    int offset = bottomJoin.getRight().getRowType().getFieldCount();
+    final ImmutableBitSet.Builder groupKey = ImmutableBitSet.builder();
+    aggregate.getGroupSet().forEach(
+        i -> groupKey.set(i >= leftBottomChildSize ? i - offset : i));
+
+    final RexBuilder rexBuilder = topJoin.getCluster().getRexBuilder();
+    final RelBuilder relBuilder = call.builder();
+
+    relBuilder.push(bottomJoin.getLeft()).push(topJoin.getRight())
+        .join(topJoin.getJoinType(),
+            RelOptUtil.transformCondition(topJoin.getCondition(), rexBuilder,
+                leftBottomChildSize, offset))
+        .aggregate(relBuilder.groupKey(groupKey.build()),
+            RelOptUtil.transformAggregateCall(aggregate.getAggCallList(),
+                leftBottomChildSize, offset));
```

Comment:
How about extract some variables for debugability and readability?

Reply:
Sure.